File: implementations/sdk/typescript/src/validation.ts
import * as Ajv2020Module from "ajv/dist/2020.js";
import * as addFormatsModule from "ajv-formats";
import type { ErrorObject, ValidateFunction } from "ajv";

const Ajv2020: any = (Ajv2020Module as any).default;
const addFormats: any = (addFormatsModule as any).default;
import { readFile } from "node:fs/promises";
import { join } from "node:path";

export type SchemaBundle = Record<string, object>;

export type SchemaBundleFile = {
  protocol_version?: string;
  schema_files?: Record<string, string>;
  schemas?: SchemaBundle;
};

export class LcpSchemaValidationError extends Error {
  constructor(public readonly schemaName: string, public readonly errors: ErrorObject[]) {
    super(`LCP schema validation failed for ${schemaName}`);
  }
}

export class LcpSchemaValidator {
  private readonly ajv: any;
  private readonly validators = new Map<string, ValidateFunction>();
  private readonly documents = new Map<string, object>();

  constructor(schemas: SchemaBundle) {
    this.ajv = new Ajv2020({ allErrors: true, strict: false, schemas: Object.values(schemas) as any });
    addFormats(this.ajv);
    for (const [name, schema] of Object.entries(schemas)) {
      const id = typeof (schema as { $id?: unknown }).$id === "string" ? (schema as { $id: string }).$id : name;
      this.documents.set(name, schema);
      this.documents.set(id, schema);
      this.validators.set(name, this.ajv.compile(schema as any));
      this.validators.set(id, this.ajv.getSchema(id) ?? this.ajv.compile(schema as any));
    }
  }

  static fromBundle(bundle: SchemaBundle | SchemaBundleFile): LcpSchemaValidator {
    const schemas = (bundle as SchemaBundleFile).schemas ?? (bundle as SchemaBundle);
    return new LcpSchemaValidator(schemas);
  }

  static async fromDirectory(root: string): Promise<LcpSchemaValidator> {
    const schemas: SchemaBundle = {};
    for (const directory of ["schemas", "verticals"]) {
      const files = await (await import("node:fs/promises")).readdir(join(root, directory));
      for (const file of files.filter((value: string) => value.endsWith(".json"))) {
        const name = `${directory}/${file}`;
        schemas[name] = JSON.parse(await readFile(join(root, directory, file), "utf8")) as object;
      }
    }
    return new LcpSchemaValidator(schemas);
  }

  validate(schemaName: string, value: unknown): void {
    const normalized = schemaName.replace(/^schemas\//, "").replace(/\.json$/, "");
    const validator = this.validators.get(schemaName) ?? this.validators.get(normalized) ?? this.validators.get(`schemas/${normalized}.json`);
    if (!validator) throw new Error(`Unknown LCP schema: ${schemaName}`);
    if (!validator(value)) throw new LcpSchemaValidationError(schemaName, validator.errors ?? []);
  }

  validateVertical(vertical: string, value: unknown): void {
    this.validate(`verticals/${vertical}.json`, value);
  }

  validateEnvelope(envelope: unknown): void {
    this.validate("schemas/envelope.json", envelope);
    const typed = envelope as { lcp: { message: { type: string }; payload: any } };
    const messageType = typed.lcp.message.type;
    this.validate(`schemas/${messageType}.json`, typed.lcp.payload);
    if (!["lead", "call", "post", "ping"].includes(messageType)) return;

    const attributes = typed.lcp.payload?.attributes;
    const vertical = messageType === "ping"
      ? typed.lcp.payload?.vertical
      : attributes?.vertical;
    if (typeof vertical !== "string" || !attributes || typeof attributes !== "object") return;

    const verticalAttributes = { ...attributes } as Record<string, unknown>;
    if (messageType === "ping") {
      verticalAttributes.vertical ??= vertical;
      const schema = this.documents.get(`verticals/${vertical}.json`) as any;
      verticalAttributes.schema_version ??= schema?.properties?.schema_version?.const ?? "1.0.0";
    }
    this.validateVertical(vertical, verticalAttributes);
    if (messageType === "ping") this.validatePingSafe(vertical, attributes);
  }

  private validatePingSafe(vertical: string, attributes: Record<string, unknown>): void {
    const schema = this.documents.get(`verticals/${vertical}.json`) as any;
    if (!schema) throw new Error(`Vertical schema '${vertical}' not found for ping-safe validation`);
    const errors: ErrorObject[] = [];
    const walk = (value: unknown, definition: any, path: string): void => {
      if (!value || typeof value !== "object" || Array.isArray(value)) return;
      const properties = definition?.properties ?? {};
      for (const [field, child] of Object.entries(value as Record<string, unknown>)) {
        if (path === "attributes" && (field === "vertical" || field === "schema_version")) continue;
        const childDefinition = properties[field] as any;
        const instancePath = `${path}.${field}`;
        if (!childDefinition || childDefinition.ping_safe !== true) {
          errors.push({
            instancePath,
            schemaPath: "#/properties/ping_safe",
            keyword: "ping_safe",
            params: {},
            message: "field is not tagged ping_safe: true",
          });
          continue;
        }
        if (childDefinition.properties) walk(child, childDefinition, instancePath);
      }
    };
    walk(attributes, schema, "attributes");
    if (errors.length) throw new LcpSchemaValidationError(`verticals/${vertical}.json`, errors);
  }
}

File: implementations/sdk/python/lcp_sdk/validation.py
"""JSON Schema validation for LCP messages."""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any

from jsonschema import Draft202012Validator
from referencing import Registry, Resource
from referencing.jsonschema import DRAFT202012


MESSAGE_TYPES = ("lead", "call", "ping", "post", "ack", "event", "bid")


def default_schema_root() -> Path:
    """Find the repository schemas when running from a source checkout."""
    configured = os.environ.get("LCP_SCHEMA_DIR")
    if configured:
        return Path(configured)
    bundled = Path(__file__).resolve().parent / "schemas"
    if bundled.exists():
        return bundled
    # implementations/sdk/python/lcp_sdk -> repository root / schemas
    return Path(__file__).resolve().parents[4] / "schemas"


class ValidationError(ValueError):
    """Raised when a message fails LCP schema validation."""

    def __init__(self, errors: list[str]):
        self.errors = errors
        super().__init__("; ".join(errors))


class SchemaValidator:
    """Validate LCP envelopes and payloads against repository schemas."""

    def __init__(self, schema_root: str | Path | None = None):
        self.schema_root = Path(schema_root) if schema_root else default_schema_root()
        self.vertical_root = self.schema_root.parent / "verticals"
        core = self._load("core")
        core_resource = Resource(contents=core, specification=DRAFT202012)
        registry = Registry().with_resource(
            uri="https://lcp.dev/schemas/core.json", resource=core_resource
        ).with_resource(uri="core.json", resource=core_resource)
        self._registry = registry
        self._schemas: dict[str, dict[str, Any]] = {
            name: self._load(name) for name in ("core", "envelope", "offer", *MESSAGE_TYPES)
        }
        self._validators = {
            name: Draft202012Validator(schema, registry=registry)
            for name, schema in self._schemas.items()
            if name != "core"
        }
        self._envelope = self._validators["envelope"]
        self._offer = self._validators["offer"]
        self._verticals: dict[str, Draft202012Validator] = {}
        if self.vertical_root.exists():
            for path in self.vertical_root.glob("*.json"):
                self._verticals[path.stem] = Draft202012Validator(
                    json.loads(path.read_text(encoding="utf-8")), registry=registry
                )

    def _load(self, name: str) -> dict[str, Any]:
        path = self.schema_root / f"{name}.json"
        with path.open(encoding="utf-8") as handle:
            return json.load(handle)

    @staticmethod
    def _format_error(error: Any) -> str:
        path = "/".join(str(part) for part in error.path)
        return f"{error.message} at {path or '(root)'}"

    @classmethod
    def from_bundle(cls, bundle: dict[str, dict[str, Any]]) -> "SchemaValidator":
        """Build a validator from the generated ``schema-bundle.json`` object."""
        instance = cls.__new__(cls)
        instance.schema_root = Path(".")
        instance.vertical_root = Path(".")
        core = bundle.get("schemas/core.json") or bundle.get("core.json")
        if core is None:
            raise ValueError("schema bundle does not contain schemas/core.json")
        resource = Resource(contents=core, specification=DRAFT202012)
        registry = Registry().with_resource(uri="https://lcp.dev/schemas/core.json", resource=resource).with_resource(uri="core.json", resource=resource)
        instance._registry = registry
        instance._schemas = {
            key.removeprefix("schemas/").removesuffix(".json"): value
            for key, value in bundle.items()
            if key.startswith("schemas/")
        }
        instance._validators = {
            name: Draft202012Validator(schema, registry=registry)
            for name, schema in instance._schemas.items()
            if name != "core"
        }
        instance._envelope = instance._validators["envelope"]
        instance._offer = instance._validators["offer"]
        instance._verticals = {
            key.removeprefix("verticals/").removesuffix(".json"): Draft202012Validator(value, registry=registry)
            for key, value in bundle.items()
            if key.startswith("verticals/")
        }
        return instance

    def validate(self, schema_name: str, value: Any) -> list[str]:
        """Validate arbitrary data against a canonical core/message/vertical schema."""
        normalized = schema_name.removeprefix("schemas/").removesuffix(".json")
        validator = self._validators.get(normalized) or self._verticals.get(
            schema_name.removeprefix("verticals/").removesuffix(".json")
        )
        if validator is None:
            raise KeyError(f"unknown LCP schema: {schema_name}")
        return [self._format_error(error) for error in validator.iter_errors(value)]

    def require_valid(self, schema_name: str, value: Any) -> None:
        errors = self.validate(schema_name, value)
        if errors:
            raise ValidationError(errors)

    def validate_envelope(self, envelope: dict[str, Any]) -> list[str]:
        """Return validation errors for an envelope and its message payload."""
        errors = [self._format_error(error) for error in self._envelope.iter_errors(envelope)]
        if errors:
            return errors
        message = envelope["lcp"]["message"]
        message_type = message["type"]
        payload = envelope["lcp"]["payload"]
        errors.extend(
            self._format_error(error)
            for error in self._validators[message_type].iter_errors(payload)
        )
        if message_type == "ping":
            errors.extend(self._ping_safe_errors(payload))
        return errors

    def require_valid_envelope(self, envelope: dict[str, Any]) -> None:
        errors = self.validate_envelope(envelope)
        if errors:
            raise ValidationError(errors)

    def validate_offer(self, offer: dict[str, Any]) -> list[str]:
        return self.validate("offer", offer)

    def require_valid_offer(self, offer: dict[str, Any]) -> None:
        errors = self.validate_offer(offer)
        if errors:
            raise ValidationError(errors)

    def _ping_safe_errors(self, payload: dict[str, Any]) -> list[str]:
        attributes = payload.get("attributes", {})
        vertical = payload.get("vertical")
        if not attributes or not vertical:
            return []
        vertical_path = self.vertical_root / f"{vertical}.json"
        if not vertical_path.exists():
            return [f"vertical schema '{vertical}' not found for ping-safe validation"]
        with vertical_path.open(encoding="utf-8") as handle:
            schema = json.load(handle)
        errors: list[str] = []
        for field_name in attributes:
            definition = schema.get("properties", {}).get(field_name, {})
            if definition.get("ping_safe") is not True:
                errors.append(
                    f"attributes.{field_name} is not tagged ping_safe: true "
                    f"in vertical '{vertical}'"
                )
        return errors
